sdk/src/types/block/output/feature/metadata.rs

```diff
@@ -27,77 +64,105 @@ impl MetadataFeature {
 
     /// Creates a new [`MetadataFeature`].
     #[inline(always)]
-    pub fn new(data: impl Into<Vec<u8>>) -> Result<Self, Error> {
-        Self::try_from(data.into())
+    pub fn new(data: impl IntoIterator<Item = (Vec<u8>, Vec<u8>)>) -> Result<Self, Error> {
+        let data: BTreeMap<Vec<u8>, Vec<u8>> = data.into_iter().collect();
+
+        for key in data.keys() {
+            if !key.iter().all(|b| b.is_ascii_graphic()) {
+                return Err(Error::NonAsciiMetadataKey(key.to_vec()));
+            }
+        }
+        Self::try_from(data)
     }
 
     /// Returns the data.
     #[inline(always)]
-    pub fn data(&self) -> &[u8] {
-        &self.0
+    pub fn data(
+        &self,
+    ) -> &BTreeMap<BoxedSlicePrefix<u8, MetadataFeatureKeyLength>, BoxedSlicePrefix<u8, MetadataFeatureValueLength>>
+    {
+        self.0.deref()
     }
 }
 
 impl StorageScore for MetadataFeature {}
 
 impl WorkScore for MetadataFeature {}
 
-macro_rules! impl_from_vec {
-    ($type:ty) => {
-        impl TryFrom<$type> for MetadataFeature {
-            type Error = Error;
+impl TryFrom<Vec<(Vec<u8>, Vec<u8>)>> for MetadataFeature {
+    type Error = Error;
 
-            fn try_from(value: $type) -> Result<Self, Self::Error> {
-                Vec::<u8>::from(value).try_into()
+    fn try_from(data: Vec<(Vec<u8>, Vec<u8>)>) -> Result<Self, Error> {
+        let mut res = BTreeMap::<
+            BoxedSlicePrefix<u8, MetadataFeatureKeyLength>,
+            BoxedSlicePrefix<u8, MetadataFeatureValueLength>,
+        >::new();
+        for (k, v) in data {
+            if !k.iter().all(|b| b.is_ascii_graphic()) {
+                return Err(Error::NonAsciiMetadataKey(k.to_vec()));
             }
+            res.insert(
```

Comment:
Should this return an error if there was already an entry?

Reply:
I would say so